usr.sbin/ctld/ctld.cc

```diff
@@ -93,231 +96,236 @@ usage(void)
 	}
+}
 
-	return (NULL);
+void
+conf::write_pidfile()
+{
+	conf_pidfile.write();
 }
 
-static void
-auth_check_secret_length(const struct auth_group *ag, const char *user,
-    const char *secret, const char *secret_type)
+void
+conf::close_pidfile()
+{
+	conf_pidfile.close();
+}
+
+#ifdef ICL_KERNEL_PROXY
+int
+conf::add_proxy_portal(portal *portal)
+{
+	conf_proxy_portals.push_back(portal);
+	return (conf_proxy_portals.size() - 1);
+}
+
+portal *
+conf::proxy_portal(int id)
+{
+	if (id >= conf_proxy_portals.size())
+		return (nullptr);
+	return (conf_proxy_portals[id]);
+}
+#endif
+
+bool
+auth_group::set_type(const char *str)
+{
+	int type;
+
+	if (strcmp(str, "none") == 0) {
+		type = AG_TYPE_NO_AUTHENTICATION;
+	} else if (strcmp(str, "deny") == 0) {
+		type = AG_TYPE_DENY;
+	} else if (strcmp(str, "chap") == 0) {
+		type = AG_TYPE_CHAP;
+	} else if (strcmp(str, "chap-mutual") == 0) {
+		type = AG_TYPE_CHAP_MUTUAL;
+	} else {
+		log_warnx("invalid auth-type \"%s\" for %s", str, label());
+		return (false);
+	}
+
+	if (ag_type != AG_TYPE_UNKNOWN && ag_type != type) {
+		log_warnx("cannot set auth-type to \"%s\" for %s; "
+		    "already has a different type", str, label());
+		return (false);
+	}
+
+	ag_type = type;
+
+	return (true);
+}
+
+void
+auth_group::set_type(int type)
+{
+	assert(ag_type == AG_TYPE_UNKNOWN);
+
+	ag_type = type;
+}
+
+const struct auth *
+auth_group::find_auth(const char *user) const
+{
+	auto it = ag_auths.find(user);
+	if (it == ag_auths.end())
+		return (nullptr);
+
+	return (&it->second);
+}
+
+void
+auth_group::check_secret_length(const char *user, const char *secret,
+    const char *secret_type)
 {
 	size_t len;
 
 	len = strlen(secret);
+	assert(len != 0);
 	if (len > 16) {
 		log_warnx("%s for user \"%s\", %s, is too long; it should be "
-		    "at most 16 characters long", secret_type, user,
-		    ag->ag_label);
+		    "at most 16 characters long", secret_type, user, label());
 	}
 	if (len < 12) {
 		log_warnx("%s for user \"%s\", %s, is too short; it should be "
-		    "at least 12 characters long", secret_type, user,
-		    ag->ag_label);
+		    "at least 12 characters long", secret_type, user, label());
 	}
 }
 
 bool
-auth_new_chap(struct auth_group *ag, const char *user,
-    const char *secret)
+auth_group::add_chap(const char *user, const char *secret)
 {
-	struct auth *auth;
-
-	if (ag->ag_type == AG_TYPE_UNKNOWN)
-		ag->ag_type = AG_TYPE_CHAP;
-	if (ag->ag_type != AG_TYPE_CHAP) {
+	if (ag_type == AG_TYPE_UNKNOWN)
+		ag_type = AG_TYPE_CHAP;
+	if (ag_type != AG_TYPE_CHAP) {
 		log_warnx("cannot mix \"chap\" authentication with "
-		    "other types for %s", ag->ag_label);
+		    "other types for %s", label());
 		return (false);
 	}
 
-	auth_check_secret_length(ag, user, secret, "secret");
+	check_secret_length(user, secret, "secret");
 
-	auth = auth_new(ag);
-	auth->a_user = checked_strdup(user);
-	auth->a_secret = checked_strdup(secret);
+	const auto &pair = ag_auths.try_emplace(user, secret);
```

Comment:
I prefer `const auto` over `auto const` (not sure why, just "looks right").  I think `auto` is also preferred in this case when you are getting a std::pair<> back (especially when the type of the pair<> is a pain.. here one template type is an iterator).